app/src/main/java/com/willowtree/vocable/presets/adapter/AddToCategoryPickerViewModel.kt

```diff
@@ -0,0 +1,154 @@
+import com.willowtree.vocable.presets.PresetsRepository
+import com.willowtree.vocable.room.Category
+import com.willowtree.vocable.room.CategoryPhraseCrossRef
+import com.willowtree.vocable.room.Phrase
+import com.willowtree.vocable.utils.LocalizedResourceUtility
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import org.koin.core.inject
+import java.util.*
+
+class AddToCategoryPickerViewModel : BaseViewModel() {
+
+    private val presetsRepository: PresetsRepository by inject()
+    private val localizedResourceUtility: LocalizedResourceUtility by inject()
+
+    private val liveCategoryMap = MutableLiveData<Map<Category, Boolean>>()
+    val categoryMap: LiveData<Map<Category, Boolean>> = liveCategoryMap
+
+    private val liveShowPhraseAdded = MutableLiveData<Boolean>()
+    val showPhraseAdded: LiveData<Boolean> = liveShowPhraseAdded
+
+    private val liveShowPhraseDeleted = MutableLiveData<Boolean>()
+    val showPhraseDeleted: LiveData<Boolean> = liveShowPhraseDeleted
+
+    private var userCategories = listOf<Category>()
+
+    // a list that maps categories to the phrase.  The phrase will not change, so we just need
+    // to map the category id to the phrase object that was added to it, so we can delete it later
+    private var savedPhrases = mutableMapOf<String, Phrase>()
+
+    fun handleCategoryToggled(phraseString: String, category: Category, isChecked: Boolean) {
+        backgroundScope.launch {
+            // if the user checked the switch
+            if (isChecked) {
+                // add it to the category
+                addNewPhrase(phraseString, category)
+            } else { // if they un-checked the switch
+                // delete it from the category
+                deletePhraseFromCategory(phraseString, category)
+            }
+        }
+    }
+
+    fun buildCategoryList(phraseString: String) {
+        // map category to whether or not the phrase is in that category
+        val phraseMap = mutableMapOf<Category, Boolean>()
+
+        backgroundScope.launch {
+            userCategories = presetsRepository.getUserGeneratedCategories()
+
+            // for each category in the list
+            userCategories.forEach { category ->
+                // get all the phrases associated with it
+                val phrases = presetsRepository.getPhrasesForCategory(category.categoryId)
+
+                // for each phrase in the category
+                phrases.forEach {
+                    // if we find a match
+                    if (localizedResourceUtility.getTextFromPhrase(it) == phraseString) {
+                        // add it to the map
+                        phraseMap[category] = true
+                    }
+                }
+
+                // if we get through the loop without adding anything, the phrase isn't in there
+                if (phraseMap[category] == null) {
+                    phraseMap[category] = false
+                }
+            }
```

Comment:
This is in a `backgroundScope` so i'm not too worried about it but do you see any bad performance issues w/ this being a nested loop? Is there any way around this? I don't have enuff context on the project but i always try to avoid nested loops as much as I can.

Reply:
I don't think there's a way to avoid this with the way the model is set up - there isn't a simple way to check if a phrase is in a category just by using the phrase string.  I also would like to avoid nested loops if possible, but in this case it's really unlikely that the user would have tons of custom categories with tons of custom phrases in each one, and since it's all in a background scope, I think it should be okay.  I'll definitely keep it in mind if we start to see performance issues though